ethereum/contracts/SimplifiedMMRVerification.sol

```diff
@@ -0,0 +1,96 @@
+// "SPDX-License-Identifier: UNLICENSED"
+pragma solidity ^0.8.5;
+
+struct SimplifiedMMRProof {
+    bytes32[] restOfThePeaks;
+    bytes32 rightBaggedPeak;
+    bytes32[] merkleProofItems;
+    uint64 merkleProofOrderBitField;
+}
+
+contract  SimplifiedMMRVerification {
+    function verifyInclusionProof(
+        bytes32 root,
+        bytes32 leafNodeHash,
+        SimplifiedMMRProof memory proof
+    ) public pure returns (bool) {
+        require(proof.merkleProofItems.length < 64);
+
+        bool hasRightBaggedPeak = proof.rightBaggedPeak != 0x0;
+
+        uint numberOfPeaks = 1 + proof.restOfThePeaks.length;
+        if (hasRightBaggedPeak) {
+            numberOfPeaks++;
+        }
+
+        bytes32[] memory reversedPeaks = new bytes32[](numberOfPeaks);
+        uint peakInsertionPointer = 0;
+
+        if (hasRightBaggedPeak) {
+            reversedPeaks[peakInsertionPointer++] = proof.rightBaggedPeak;
+        }
+
+        bytes32 merkleRootPeak = calculateMerkleRoot(leafNodeHash, proof.merkleProofItems, proof.merkleProofOrderBitField);
+        reversedPeaks[peakInsertionPointer++] = merkleRootPeak;
+
+        if (proof.restOfThePeaks.length > 0) {
+            for (uint i = 0; i < proof.restOfThePeaks.length; i++) {
+                reversedPeaks[peakInsertionPointer] = proof.restOfThePeaks[proof.restOfThePeaks.length - i - 1];
+                peakInsertionPointer++;
+            }
+        }
```

Comment:
I still don't understand why we cannot convert this whole logic into a simple merkle proof. I made an example that shows how a MMR can be transformed into a simple merkle proof – could you point out where I am wrong or provide a counterexample? 

<img width="876" alt="Screenshot 2021-09-16 at 23 08 52" src="https://user-images.githubusercontent.com/6912756/133686013-404312d2-532e-4e0a-8d4d-55f35fa1c841.png">

Thanks!

Reply:
Part of the reason is the nodes just below the root have different handling then in binary merkle proof. In binary merkle tree those nodes can only be two, so the merkle proof generation/verification code can handle them easily using the recursive algorithm. 

But in MMR the nodes just below root can be any number and the algorithm to calculate root is different.

Reply:
Could you give me an example where the peaks are not handled in a binary way? 

The code you implemented does pairwise (binary) hashing: https://github.com/Snowfork/snowbridge/pull/495/files/ec5a2a4cccf4e17998a9dacf1ac6ba2f3c9308ab#diff-ccff880ac8ad6b758b2b22c2c6ed830cb524733bd4a05d38d3ad37165f6333e1R55 
Also the example you give `Hash(B, Hash(C, D))` is pairwise, and can be represented as a binary merkle tree.

Do you have an example where the parent is composed of more than two children?

Reply:
Here is a graphic for how the MMR above looks as a binary merkle tree: 

<img width="325" alt="Screenshot 2021-09-17 at 08 43 04" src="https://user-images.githubusercontent.com/6912756/133736732-916c3b83-e5fe-47dd-bed5-b824331936dc.png">

Reply:
Ok. Simplified the calculation.